app/src/main/java/org/p2p/wallet/history/interactor/TransactionDetailsMapper.kt

```diff
@@ -0,0 +1,258 @@
+import org.p2p.wallet.history.db.entities.RenBtcBurnOrMintTransactionEntity
+import org.p2p.wallet.history.db.entities.SwapAEntity
+import org.p2p.wallet.history.db.entities.SwapBEntity
+import org.p2p.wallet.history.db.entities.SwapTransactionEntity
+import org.p2p.wallet.history.db.entities.TransactionEntity
+import org.p2p.wallet.history.db.entities.TransferTransactionEntity
+import org.p2p.wallet.history.db.entities.UnknownTransactionEntity
+import org.p2p.wallet.history.db.entities.embedded.CommonTransactionInformationEntity
+import org.p2p.wallet.history.db.entities.embedded.TransactionIdentifiersEntity
+import org.p2p.wallet.history.db.entities.embedded.TransactionTypeEntity
+import org.p2p.wallet.utils.fromJsonReified
+import org.p2p.wallet.utils.toBase58Instance
+import timber.log.Timber
+
+class TransactionDetailsMapper(
+    private val confirmedTransactionParser: ConfirmedTransactionParser
+) {
+    fun mapDtoToDomain(confirmedTransactions: List<ConfirmedTransactionParsed>): List<TransactionDetails> {
+        val resultTransactions = mutableListOf<TransactionDetails>()
+
+        confirmedTransactions.forEach { confirmedTransaction ->
+            val parsedTransactions = confirmedTransactionParser.parseToTransactionDetails(confirmedTransaction)
+
+            val swapTransaction = parsedTransactions.firstOrNull { it is SwapDetails }
+            if (swapTransaction != null) {
+                resultTransactions.add(swapTransaction)
+                return@forEach
+            }
+
+            val burnOrMintTransaction = parsedTransactions.firstOrNull { it is BurnOrMintDetails }
+            if (burnOrMintTransaction != null) {
+                resultTransactions.add(burnOrMintTransaction)
+                return@forEach
+            }
+
+            val transferTransaction = parsedTransactions.firstOrNull { it is TransferDetails }
+            if (transferTransaction != null) {
+                resultTransactions.add(transferTransaction)
+                return@forEach
+            }
+
+            val createTransaction = parsedTransactions.firstOrNull { it is CreateAccountDetails }
+            if (createTransaction != null) {
+                resultTransactions.add(createTransaction)
+                return@forEach
+            }
+
+            val closeTransaction = parsedTransactions.firstOrNull { it is CloseAccountDetails }
+            if (closeTransaction != null) {
+                resultTransactions.add(closeTransaction)
+                return@forEach
+            }
+
+            val unknownTransaction = parsedTransactions.firstOrNull { it is UnknownDetails }
+            if (unknownTransaction != null) {
+                resultTransactions.add(unknownTransaction)
+                return@forEach
+            }
+        }
+
+        return resultTransactions.toList()
+    }
+
+    fun mapEntityToDomain(entities: List<TransactionEntity>): List<TransactionDetails> {
```

Comment:
My goal was to preserve all the mapping and init logic that are still in Java POJO TransactionDetails.
so safe mapping is goes like this:
networkPojo -> TransactionDetails -> TransactionEntity -> TransactionDetails (there is some strange mapping going on in constructors, want to save that logic) -> HistoryTransaction

Reply:
Maybe, it would be better to refactor java classes to kotlin and remove redundant mapping step? 
But this task would be pretty large so, let's do in a separate ticket